ourhexenv.py

```diff
@@ -166,21 +166,23 @@ def reset(self):
         if self.window:
             self.window.fill(self.BACKGROUND)
             pygame.display.flip()
-        
 
     def step(self, action):
+        #Check if the action is within the valid range.
+        if action not in range(self.action_space.n):
+            raise ValueError("Illegal move: Action is out of bounds.")
         # Handle pie rule
         if action == self.board_size * self.board_size:
             if self.agent_selection == "player_1":
                 raise ValueError("Illegal move: Pie rule can only be used by Player 2.")
             if not self.is_pie_rule_usable:
                 raise ValueError("Illegal move: Pie rule can only be used once.")
 
-            self.is_pie_rule_usable = True
-
-            self.board = np.where(self.board == 1, 3, self.board)
-            self.board = np.where(self.board == 2, 1, self.board)
-            self.board = np.where(self.board == 3, 2, self.board)
+            # Use pie rule, if a (row,col) was 1, make it 0 and make (col,row) 1
```

Comment:
I've seen two variations of the pie rule in different literature about the game.

1. Mirroring: In this approach the pie rule allows the second player to remove the first player's tile and place their own on the position reflected across the diagonal of the board. This is implementation we are looking at in this PR.
2. Swapping: In this approach, the second player claims the first player's tile as his own and the two switch colors (and corresponding board end zones). This is the approach you're discussing.

I think we are going with the first approach (mirroring) because we agreed in class that we don't want to add additional logic to switch the colors (feel free to correct me here because I wasn't paying much attention last week).

Lmk if something doesn't make sense.